### DISCUSSION TITLE
Bug: last part of 'command' is being truncating causing steady error

### DISCUSSION THREAD

**Discussion by Danylov-Mykola** [woodpecker-ci/woodpecker#3973] [Ask a Question] [answered]
There is the .woodpecker.yaml part for a 'commands' section is hardly limited (!?). That looks like a **serious bug** that could not be visible with a simple pipeline.
![woodpecker-error](https://github.com/user-attachments/assets/260a5cfd-14e5-464a-bb71-e3d55d533541)
During a build all exceeded characters of a 'step' are eliminated, causing a runtime syntax error of a shell command execution. There is only workaround is split the commands into several steps. Unfortunately in commands such as 'docker service create' we need to use tens of parameters with a single shell command.
So, is there another way to perform "long" steps normally?
How the bug could be reported for fixing?

**Comment by qwerty287** [comment]
Can you post your config?

**Reply by Danylov-Mykola** [reply]
Thanks.
The config with additional information is in created issue. I've post there additional information.
See: https://github.com/woodpecker-ci/woodpecker/issues/3974

**Comment by zc-devs** [comment]
> So, is there another way to perform "long" steps normally?

Make the scripts and run them from the commands.
For example, `create-service.sh`:
```
#!/bin/sh

ssh -o "StrictHostKeyChecking no" $CI_USER_NAME@$DEPLOY_HOSTIP
        "docker service create
         ...
        "
```
```yaml
steps:
  create-or-update-service:
    commands:
      - .set-env.sh
      - .pull-image.sh
      - .create-service.sh
      - ...
```

**Reply by Danylov-Mykola** [reply]
Thank you, that helps a litttle. We already use the specified method. However, the indicated errors still have to be bypassed additionally, shortening the "commands" section as well. And suppressing the error of exceeding the limit adds complexity and needs to always pay attention. So, now we need to pass the parameters from the "woodpecker" now to the script, and the number of parameters is the same.
Note that we want to separate the parameters from the pipeline "code" because the woodpecker config should remain the same for different environments. So we will have:
```
- ssh -o "StrictHostKeyChecking no "$$CI_USER_NAME@$$DEPLOY_HOSTIP "chmod u+x ./create-service.sh && ./create-service.sh -i=$$IMAGE_NAME -n=$$SWARM_DOCKER_NETWORK"
```
instead of:
```
- ssh -o "StrictHostKeyChecking no "$$CI_USER_NAME@$$DEPLOY_HOSTIP "docker service create --network $$SWARM_DOCKER_NETWORK $$IMAGE_NAME"
```
As you can see, this method does not help shorten the command much, unless the parameters are **hardcoded** inside the script.

**Reply by zc-devs** [reply]
> chmod u+x

You can commit & push it to the repo
```
git update-index --chmod=+x .ci/*.sh
```

> we need to pass the parameters from the "woodpecker" now to the script

You can use [env variables, imports](https://gitea.smthd.com/mirko/radicale-docker), [libraries and so on](https://github.com/bitnami/containers/tree/main/bitnami/mongodb/7.0/debian-12/prebuildfs/opt/bitnami/scripts). In the end, it all comes down to your Shell-guru skills.

**Comment by qwerty287** [comment]
@zc-devs's solution is possible, but yaml actually supports multiline strings itself: https://yaml-multiline.info/

You can just use that and add a `\` to the end of each line so your shell knows the command isn't done yet.

**Reply by Danylov-Mykola** [reply]
I understand where your point is coming from. Unfortunately, all of leading spaces from .woodpecker.yaml are goes to preprocessing and taking in the consider increasing a "command" string inside woodpecker app. That's why some commands in configuration we have to use with a single and hardly readable string.

**Comment by Danylov-Mykola** [comment]
@qwerty287, it is sad to know that the config is wrong as you mentioned there https://github.com/woodpecker-ci/woodpecker/issues/3974#issuecomment-2250574877 . That means there is no way to fix it...
Each time I have the described error with:
```
    commands:
      - ...
      - shell_command1 -parameter1 value1 -parameter2 -value2
      - shell_command2 -parameter1 value1 -parameter2 -value2
```
and replacing it with:
```
    commands:
      - ...
      - shell_command1 -parameter1 value1
      - shell_command2 -parameter1 value1 -parameter2 -value2
```
makes a miracle.
Do we prohibited to use multiline shell commands with quoted parts?

**Comment by qwerty287** [comment, marked as answer]
> Do we prohibited to use multiline shell commands with quoted parts?

If we mean the syntax you're using currently, yes, this is prohibited by the YAML specs.

> Unfortunately, all of leading spaces from .woodpecker.yaml are goes to preprocessing and taking in the consider increasing a "command" string inside woodpecker app

I don't get that.
YAML has support for multiline strings and the . Please try something like this:
```yaml
steps:
 - name: xyz
   commands:
    - |
     ssh-whatever with a "long
     multiline
     arg"
```

This should parse as
```
ssh-whatever with a "long
multiline
arg"
```

If you want the newlines in there replaced by spaces, just replace the `|` by `>`.

---

And, in the future, please don't be that placative like "That means there is no way to fix it...".
It's totally fine to ask things here, but not everything is a bug.

**Reply by Danylov-Mykola** [reply]
Wow! @qwerty287 at the end you have made my day!
Indeed, the next format you offered **works well** without such strange length-dependent error:
```
    commands:
      - |
        ssh -o "StrictHostKeyChecking no" $$CI_USER_NAME@$$DEPLOY_HOSTIP \
        "docker service create \
        --constraint=\"node.role==manager\" \
        --mode=global \
        ..."
```
_*) using a '>+' looks pretty well also._
but the next one, we used to use instead is _working_ without errors **just for short commands** part:

```
    commands:
      - ssh -o "StrictHostKeyChecking no" $$CI_USER_NAME@$$DEPLOY_HOSTIP 
        "docker service create 
        --constraint=\"node.role==manager\" 
        --mode=global 
        ..."
```
I apologize for the extra trouble I have caused with my issue. It's worth saying that I'm on _your_ side and I wish your project a long and wonderful history.
**Thank you again for all you do!**
